File: Sorting/CocktailShakerSort/cocktail_shaker_sort.js
function swap(array) { // create a swap function specific to an array
  return function (x, y) {
    var value = array[x];
    array[x] = array[y];
    array[y] = value;
  };
}

function cocktail_shaker_sort(array) {
  var _swap = swap(array); // when given x and y swap elements in this array
  var swapped = false;
  var last = array.length - 1;
  var first = 0;
  do {
    for (var i = first; i < last; i++) {
      if (array[i] > array[i + 1]) { // test whether the two elements are in the wrong order
        _swap(i, i + 1); // let the two elements change places
        swapped = true;
      }
    }
    if (!swapped) break; // we can exit the outer loop here if no swaps occurred.
    swapped = false;
    last--; // the last element is already sorted
    for (var i = last; i > first; i--) {
      if (array[i] < array[i - 1]) { // test whether the two elements are in the wrong order
        _swap(i, i - 1); // let the two elements change places
        swapped = true;
      }
    }
    first++; // the first element is already sorted
  } while (swapped); // if no elements have been swapped, then the list is sorted
}

File: Sorting/Cocktail Shaker Sort/Python/Cocktail_Shaker_Sort.py
#Cocktail Shaker Sort function
def cocktail_shaker_sort(array):
    swapped = True

    #Initialize first element to 0 and last element to n-1
    first = 0
    last = len(array) - 1

    while swapped:
        swapped = False

        #Forward pass
        for i in range(first,last):
            if array[i] > array[i+1]:
                array[i], array[i+1] = array[i+1], array[i]
                swapped = True

        #If there are no swaps, then array is sorted
        if not swapped:
            break

        last -= 1
        swapped = False

        #Backward pass
        for i in range(last,first,-1):
            if array[i] < array[i-1]:
                array[i], array[i-1] = array[i-1], array[i]
                swapped = True

        first += 1

if __name__ == "__main__":
    array = [1,2,9,1,2,10,6,3,6,8,0]
    print(array)
    cocktail_shaker_sort(array)
    print(array)
